src/consumer.js

```diff
@@ -0,0 +1,229 @@
+      this.consumerTag = null;
+      await channel.cancel(consumerTag);
+    }
+
+    // if messages are being processed, arrange to continue when they
+    // are all handled
+    if (this.processingMessages > 0) {
+      await new Promise(resolved => {
+        this.idleCallback = resolved;
+      });
+    }
+
+    if (channel) {
+      this.channel = null;
+      await channel.close();
+    }
+  }
+
+  /**
+   * Handle a new connection to the pulse server, re-declaring everything
+   * and setting up the consumer.
+   */
+  async _handleConnection(conn) {
+    try {
+      if (!this.running) {
+        return;
+      }
+
+      const amqp = conn.amqp;
+      const channel = await amqp.createChannel();
+      await channel.prefetch(this.options.prefetch);
+
+      this.channel = channel;
+
+      const exclusive = !this.options.queueName;
+      const queueName = this.client.fullObjectName(
+        'queue',
+        this.options.queueName || `exclusive/${slugid.v4()}`);
+
+      await channel.assertQueue(queueName, {
+        exclusive: exclusive,
+        durable: !exclusive,
+        autoDelete: exclusive,
+        maxLength: this.options.maxLength,
+      });
+
+      for (let {exchange, routingKeyPattern} of this.bindings) {
+        await channel.bindQueue(queueName, exchange, routingKeyPattern);
+      }
+
+      this._queueBound && this._queueBound(); // (hook for tests)
+
+      const consumer = await channel.consume(queueName, async (msg) => {
+        try {
+          this.processingMessages++;
+          try {
+            await this._handleMessage(msg);
+          } catch (err) {
+            if (msg.fields.redelivered) {
+              // if this was already delivered, we're going to give up and report it
+              channel.nack(msg, false, false);
+              this.client.monitor.reportError(err, {
+                queueName,
+                exchange: msg.exchange,
+                redelivered: msg.redelivered,
+              });
+            } else {
+              channel.nack(msg, false, true);
+            }
+            return;
+          }
+          channel.ack(msg);
+        } catch (err) {
+          // the error handling in the inner try block went badly, so this
+          // channel is probably sick
+          this.client.monitor.reportError(err, {
+            queueName,
+            exchange: msg.exchange,
+          });
+          conn.failed();
+        } finally {
+          this.processingMessages--;
+          if (this.processingMessages === 0 && this.idleCallback) {
+            this.idleCallback();
+          }
+        }
+      });
+      this.consumerTag = consumer.consumerTag;
+
+      // when retirement of this connection begins, stop consuming on this
+      // channel and close the channel as soon sa all messages are handled.
+      conn.on('retiring', () => this._shutdown());
+    } catch (err) {
+      this.client.monitor.reportError(err);
+      conn.failed();
```

Comment:
So binding to an exchange that doesn't exist breaks the entire connection, and not just the channel?

Reply:
Well, it causes the connection to be restarted.  There's no harm in that (and it's not a common case anyway).